spec/services/catalog/copy_portfolio_item_spec.rb

```diff
@@ -0,0 +1,64 @@
+describe Catalog::CopyPortfolioItem do
+  let(:tenant) { create(:tenant) }
+  let(:portfolio) { create(:portfolio, :tenant_id => tenant.id) }
+  let(:portfolio2) { create(:portfolio, :tenant_id => tenant.id) }
+  let(:portfolio_item) { create(:portfolio_item, :portfolio_id => portfolio.id, :tenant_id => tenant.id) }
+
+  let(:copy_portfolio_item) { described_class.new(params).process }
+
+  describe "#process" do
+    context "when copying into the same portfolio" do
+      around do |example|
+        bypass_rbac { example.call }
+      end
+
+      let(:params) { { :portfolio_item_id => portfolio_item.id, :portfolio_id => portfolio.id } }
+
+      it "makes a copy of the portfolio_item" do
+        new = copy_portfolio_item.new_portfolio_item
+
+        expect(new.description).to eq portfolio_item.description
+        expect(new.owner).to eq portfolio_item.owner
+      end
+
+      it "modifies the name with 'Copy of'" do
+        expect(copy_portfolio_item.new_portfolio_item.name).to match(/^Copy of.*/)
+      end
+    end
+
+    context "when copying into a different portfolio" do
+      around do |example|
+        bypass_rbac { example.call }
+      end
+
+      let(:params) { { :portfolio_item_id => portfolio_item.id, :portfolio_id => portfolio2.id } }
+
+      it "makes a complete copy of the portfolio_item" do
+        new = copy_portfolio_item.new_portfolio_item
+
+        expect(new.description).to eq portfolio_item.description
+        expect(new.owner).to eq portfolio_item.owner
+        expect(copy_portfolio_item.new_portfolio_item.name).to eq portfolio_item.name
+      end
+    end
+  end
+
+  describe "#get_index" do
```

Comment:
I don't think we should be testing a private method. You should control the flow of the program via the parameters you pass in and test it in the `#process` method. If that makes the test too unwieldy, then you have a candidate for breaking out this method (and potentially others) into a separate class/module/whatever that you can then test in isolation.

So for these tests you basically need to make sure the new portfolio item that gets created via `#process` has the right names, and you'll likely need to create a few more portfolio items to test the various scenarios of "Copy of" vs "Copy 2 of". To me, again, this seems like maybe something that could be broken out into another class (`CopyNameFigureOuter` is a good name, eh? 😄).

You're also missing a spec for when the `params` actually contain a `name` that should be used.

Reply:
Broke it out into a lib module like I said earlier.

I also do have a spec for the specified name in the portfolio_item requests spec: 
https://github.com/ManageIQ/catalog-api/pull/289/files#diff-9b196a6baa076bf2384fab77223ec7feR264